File: src/main/java/com/secuworm/endpointcollector/application/ExportService.java
package com.secuworm.endpointcollector.application;

import com.secuworm.endpointcollector.domain.EndpointRecord;

import java.io.BufferedWriter;
import java.io.FileWriter;
import java.io.IOException;
import java.util.List;

public class ExportService {
    public static final String[] HEADER = new String[]{"Endpoint", "Host", "SourceURL", "Count"};

    public ExportResult exportCsv(String filePath, List<EndpointRecord> records) {
        if (filePath == null || filePath.trim().isEmpty()) {
            return new ExportResult(false, "File path is empty.");
        }

        try (BufferedWriter writer = new BufferedWriter(new FileWriter(filePath))) {
            writer.write(toCsvLine(HEADER));
            writer.newLine();
            if (records != null) {
                for (EndpointRecord record : records) {
                    if (record == null) {
                        continue;
                    }
                    String[] row = new String[]{
                        safe(record.getEndpointUrl()),
                        safe(record.getHost()),
                        safe(record.getSourceUrl()),
                        String.valueOf(record.getCount())
                    };
                    writer.write(toCsvLine(row));
                    writer.newLine();
                }
            }
            return new ExportResult(true, "");
        } catch (IOException ex) {
            return new ExportResult(false, ex.getMessage());
        }
    }

    private String safe(String value) {
        return value == null ? "" : value;
    }

    private String toCsvLine(String[] values) {
        StringBuilder builder = new StringBuilder();
        for (int i = 0; i < values.length; i++) {
            if (i > 0) {
                builder.append(',');
            }
            builder.append(escape(values[i]));
        }
        return builder.toString();
    }

    private String escape(String value) {
        String text = value == null ? "" : value;
        boolean needsQuote = text.contains(",") || text.contains("\"") || text.contains("\n") || text.contains("\r");
        String escaped = text.replace("\"", "\"\"");
        if (needsQuote) {
            return "\"" + escaped + "\"";
        }
        return escaped;
    }

    public static class ExportResult {
        private final boolean success;
        private final String errorMessage;

        public ExportResult(boolean success, String errorMessage) {
            this.success = success;
            this.errorMessage = errorMessage;
        }

        public boolean isSuccess() {
            return success;
        }

        public String getErrorMessage() {
            return errorMessage;
        }
    }
}

File: src/application/export_service.py
import csv
import sys


class ExportService(object):
    HEADER = ["Endpoint", "Host", "SourceURL", "Count"]
    IS_PY2 = sys.version_info[0] == 2

    def export_csv(self, file_path, records):
        if file_path is None or str(file_path).strip() == "":
            return False, "File path is empty."

        try:
            rows = self._to_rows(records)
            if self.IS_PY2:
                file_handle = open(file_path, "wb")
            else:
                file_handle = open(file_path, "w", newline="")
            try:
                writer = csv.writer(file_handle)
                writer.writerow(self.HEADER)
                for row in rows:
                    writer.writerow(row)
            finally:
                file_handle.close()
            return True, ""
        except Exception as exc:
            return False, str(exc)

    def _to_rows(self, records):
        rows = []
        for record in records or []:
            endpoint_url = self._read_value(record, "endpoint_url")
            host = self._read_value(record, "host")
            source_url = self._read_value(record, "source_url")
            count = self._read_value(record, "count")
            rows.append([endpoint_url, host, source_url, count])
        return rows

    def _read_value(self, record, key):
        if record is None:
            return ""
        if isinstance(record, dict):
            value = record.get(key, "")
        else:
            value = getattr(record, key, "")
        if value is None:
            return ""
        return value
